src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.cs

```diff
@@ -868,7 +869,12 @@ public string ToHexString()
             if (_asHexString == null)
             {
                 fixed (ulong* idPtr = &_id1)
-                    _asHexString = SpanToHexString(new ReadOnlySpan<byte>(idPtr, sizeof(ulong) * 2));
+                {
+                    // Cast away the read-only-ness of _asHexString, and assign the converted value to it.  
+                    // We are OK with this because conceptually the class is still read-only.  
+                    ref string strRef = ref Unsafe.AsRef(in _asHexString);
+                    strRef = SpanToHexString(new ReadOnlySpan<byte>(idPtr, sizeof(ulong) * 2));
```

Comment:
I don't have a strong opinion here, but should we consider using an Interlocked.CompareExchange to write this?  I'm just thinking that if we're trying to maintain the illusion of this truly being readonly, we'd want to do so for reference identity, too, even in the face of races.